### PR TITLE
Numpyro MCMC autodims migration

### PR DESCRIPTION
**Pull request by kylejcaron**
## Description

Part of: https://github.com/arviz-devs/arviz/issues/2453 

This is the first step, migrating the MCMC dim inferring behavior from [`arviz`](https://github.com/arviz-devs/arviz/pull/2441), before implementing the same tooling for svi models.

### The behavior
If dims are not provided, `az.from_numpyro` now infers dims automatically based on model plates, sample sites, and provided `coords`. 

For event_dims...
 * for numpyro.sample sites, infer={"event_dims":["dim1"]} can be provided
 * for deterministic sites, there is an `extra_event_dims` field in az.from_numpyro to manually provide them.

`pred_dims` are also now inferred by default if they arent provided.

### CODE REVIEW

**Comment by kylejcaron** [comment]
Any ideas why the numpyro external tests dont seem to be running in the [pipeline](https://dev.azure.com/ArviZ/ArviZ/_build/results?buildId=7234&view=logs&j=c544526c-e162-5859-cdf3-3593111709b0&t=766ad06c-c2fb-56c8-cb3f-df4b6cff2b9f). It looks like it [installed](https://dev.azure.com/ArviZ/ArviZ/_build/results?buildId=7234&view=logs&j=c544526c-e162-5859-cdf3-3593111709b0&t=0516e05b-fc9f-51dd-cbbb-3806e8b8d838) just fine

File: src/arviz_base/io_numpyro.py [hunk 1]
```diff
@@ -335,10 +454,23 @@ def from_numpyro(
...
+    If no dims are provided, this will infer batch dim names from NumPyro model plates.
+    For event dim names, such as with the ZeroSumNormal, `infer={"event_dims":dim_names}`
+    can be provided in numpyro.sample, i.e.::
+        # equivalent to dims entry, {"gamma": ["groups"]}
+        gamma = numpyro.sample(
+            "gamma",
+            dist.ZeroSumNormal(1, event_shape=(n_groups,)),
+            infer={"event_dims":["groups"]}
+        )
+    There is also an additional `extra_event_dims` input to cover any edge cases, for instance
```
Comment (OriolAbril) [inline src/arviz_base/io_numpyro.py:471]: This seems to be rendering correctly but sphinx is complaining: https://app.readthedocs.org/projects/arviz-base/builds/29253126/#282835299--66. The [official rule](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#literal-blocks) is both indented and separated from neighbouring paragraphs by blank lines

```suggestion
    can be provided in numpyro.sample, i.e.::

        # equivalent to dims entry, {"gamma": ["groups"]}
        gamma = numpyro.sample(
            "gamma",
            dist.ZeroSumNormal(1, event_shape=(n_groups,)),
            infer={"event_dims":["groups"]}
        )

    There is also an additional `extra_event_dims` input to cover any edge cases, for instance
```

File: external_tests/test_numpyro.py [hunk 2]
```diff
@@ -273,3 +281,121 @@ def model():
...
+            sigma = numpyro.sample("sigma", dist.HalfNormal(1))
+            with numpyro.plate("obs_idx", 3):
+                # mu is plated by obs_idx, but isnt broadcasted to the plate shape
+                # the expected behavior is that this should cause a failure
+                mu = numpyro.deterministic("mu", alpha)
+                return numpyro.sample("y", dist.Normal(mu, sigma), obs=jnp.array([-1, 0, 1]))
+
+        mcmc = MCMC(NUTS(model), num_warmup=10, num_samples=10)
+        mcmc.run(PRNGKey(0))
+        inference_data = from_numpyro(mcmc, coords={"obs_idx": np.arange(3)})
```
Comment (OriolAbril) [inline external_tests/test_numpyro.py:373]: will the error be when executing this function? Or are the asserts the ones that fail? If possible I would prefer using a `with pytest.raises()` if error and switching the assertions otherwise. Trying to be a bit more strict with tests than in current arviz

**Review by OriolAbril — COMMENTED** [review]
test failures are due to https://github.com/arviz-devs/arviz-base/issues/90 so don't worry about them

File: src/arviz_base/io_numpyro.py (hunk 1 above)
Comment (kylejcaron) [reply src/arviz_base/io_numpyro.py:471]: just fixed!

File: external_tests/test_numpyro.py (hunk 2 above)
Comment (kylejcaron) [reply external_tests/test_numpyro.py:373]: nice catch, it shouldnt be like that, it should just stop at from_numpyro. updated!

File: src/arviz_base/io_numpyro.py [hunk 3]
```diff
@@ -10,6 +13,88 @@
...
+        A numpyro model
+    model_args : tuple, optional
+        Input args for the numpyro model.
+    model_kwargs : dict, optional
+        Input kwargs for the numpyro model.
+
+    Returns
+    -------
+    dict[str, list[str]]
+    """
```
[the remark this answers is not in the document]
Comment (OriolAbril) [reply src/arviz_base/io_numpyro.py]: this might be worth an issue on docstub to discuss the behaviour, on one hand I think it is a bad default, but fixing it would mean singling out `tuple` out of the other containers like `list`, `sequence`, `iterable`... maybe there are other similar objects (even if not builtins) that should also be singled out along `tuple`.

As a workaround, we only do `*model_args` with this argument, so it isn't really restricted to being a tuple or even to immutable containers, so we could update the type hint to `iterable of any`. I think we already have an alias for [Iterable](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes) but not completely sure as we usually need `sequence` to use `len`, `getitem` or `contains`; either way it won't hurt to add the alias

File: src/arviz_base/io_numpyro.py [hunk 4]
```diff
@@ -10,6 +11,104 @@
...
+        Mapping from model site name to list of dimension labels.
+    """
+    import jax
+    from numpyro import distributions as dist
+    from numpyro import handlers
+    from numpyro.infer.initialization import init_to_sample
+    from numpyro.ops.pytree import PytreeTrace
+
+    model_args = tuple() if model_args is None else model_args
+    model_kwargs = dict() if model_args is None else model_kwargs
```
Comment (OriolAbril) [inline src/arviz_base/io_numpyro.py:70]:
```suggestion
    model_kwargs = dict() if model_kwargs is None else model_kwargs
```

**Review by OriolAbril — COMMENTED** [review]
I found a typo by chance and answered on the type hints thread. Let me know which is your preferred way to go about improving the type hint and I'll merge

File: src/arviz_base/io_numpyro.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kylejcaron) [reply src/arviz_base/io_numpyro.py]: Found a solution for now with their other recommended way to define [variable length tuples](https://github.com/scientific-python/docstub/blob/7fce493ada2d8308fcc3ce1a8ac49383d208dc4b/docs/typing_syntax.md?plain=1#L58) so I think we should be good to go on this now!

I did think tuple of Any(s) would work [given this](https://github.com/scientific-python/docstub/blob/7fce493ada2d8308fcc3ce1a8ac49383d208dc4b/docs/typing_syntax.md?plain=1#L51) so I'll raise an issue in docstub as well
